### ISSUE TITLE
DEEPSLEEP_RESET gets stuck after a few resets

### ISSUE THREAD

**Issue by jh83** [espressif/arduino-esp32#796]
### Hardware:
Board:							ESP32 Dev Module - Sparkfun ESP32 Thing
Core Installation/update date:			30/oct/2017
IDE name:							Arduino IDE
Flash Frequency:					80Mhz
Upload Speed:						921600

### Description:
The board does not correctly restart after deep sleep. Im using the builtin example sketch "TimerWakeUp" without modifications and no other hardware attached and i get random results on the powerup after sleep.

The board usually start successfully a few times after deepsleep, but then it failes to run the actual sketch after startup and a manual reset is needed.

The debug information below shows that the device starts after deepsleep and outputs "Boot number: x", but then in the last boot it doesn't output anything from the sketch and does not return to sleep. A manual press of the reset button restarts the device and the it runs for a few more deepsleep cycles before it gets stuck again.

Any one else seen this problem?

### Sketch:
builtin "TimerWakeUp" example sketch

### Debug Messages:
```
rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0018,len:4
load:0x3fff001c,len:956
load:0x40078000,len:0
load:0x40078000,len:11904
entry 0x40078a3c
Boot number: 1
Wakeup was not caused by deep sleep
Setup ESP32 to sleep for every 5 Seconds
Going to sleep now
ets Jun  8 2016 00:22:57

rst:0x5 (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0018,len:4
load:0x3fff001c,len:956
load:0x40078000,len:0
load:0x40078000,len:11904
entry 0x40078a3c
Boot number: 2
Wakeup caused by timer
Setup ESP32 to sleep for every 5 Seconds
Going to sleep now
ets Jun  8 2016 00:22:57

rst:0x5 (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0018,len:4
load:0x3fff001c,len:956
load:0x40078000,len:0
load:0x40078000,len:11904
entry 0x40078a3c

```

**Comment by iot-man** [comment]
I'm having the same problem. My sketch goes to sleep for 60 seconds after taking temp/humidity readings etc via I2C. It used to work ok before system upgraded to most recent commit. First two restarts work ok but always hangs on third restart.

**Debug messages**

`rst:0x5` (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0018,len:4
load:0x3fff001c,len:956
load:0x40078000,len:0
load:0x40078000,len:11904
entry 0x40078a3c
rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0018,len:4
load:0x3fff001c,len:956
load:0x40078000,len:0
load:0x40078000,len:11904
entry 0x40078a3c
Boot number: 1
Setup ESP32 to sleep for every 60 Seconds

Connecting to SKYE3E0B
MQTT connecting ...connected
19.74
1015.97
Going to sleep now
ets Jun  8 2016 00:22:57

rst:0x5 (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0018,len:4
load:0x3fff001c,len:956
load:0x40078000,len:0
load:0x40078000,len:11904
entry 0x40078a3c
Boot number: 2
Setup ESP32 to sleep for every 60 Seconds

Connecting to SKYE3E0B
MQTT connecting ...connected
19.73
1015.91
Going to sleep now
ets Jun  8 2016 00:22:57

rst:0x5 (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0018,len:4
load:0x3fff001c,len:956
load:0x40078000,len:0
load:0x40078000,len:11904
entry 0x40078a3c

**Comment by weizenbock** [comment]
I'm having a "similar" problem.  In my case it happened with ESP.restart().
It gets successfully restarted a few times, but after several cycles it hangs like below.

```
rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0018,len:4
load:0x3fff001c,len:956
load:0x40078000,len:0
load:0x40078000,len:11904
entry 0x40078a3c
```

**Comment by copercini** [comment]
Same here, after commit https://github.com/espressif/arduino-esp32/commit/55289a45af1e7898ca0b8da037d92a83baaf0083 this problems starts

**Comment by degehi** [comment]
Same here, hangs after 2-3 reboot, just after the 'entry' log line

**Comment by etimou** [comment]
Exactly the same problem for me as well with an unmodified "TimerWakeUp" example sketch.

**Comment by me-no-dev** [comment]
I would suggest an issue to be opened in IDF (if one does not already exist). It is surely something in that update.

**Comment by mikey60** [comment]
I reset the core version back to  https://github.com/espressif/arduino-esp32/tree/95123681d5340defa24472e441758fba88e3a34f  as was suggested in Issue #801 and my sketch with a 20 minute deep sleep that would hang after an hour or so (as described above) now has been working properly for 2 days.

**Comment by cristianhumelnicu** [comment]
Any patch for deepsleep issue ???

**Comment by etimou** [comment]
I can do the test with older version as well, but I am not very familiar
with git.
I guess the following commands on debian must be adapted:

git clone https://github.com/espressif/arduino-esp32.git esp32

git submodule update --init --recursive

What is the way to follow?


On Wed, Nov 8, 2017 at 10:28 AM, cristianhumelnicu <notifications@github.com

**Comment by degehi** [comment]
@etimou: what I did: just apply a checkout after de clone operation, with a version number before 55289a4

on the esp32 directory: git checkout 9512368

running without reboot hangs with this release for me.

**Comment by etimou** [comment]
thank you! it works perfectly with this version.

**Comment by cristianhumelnicu** [comment]
@etimou, @degehi 
Could you tell me what is the maximum time you've been able to keep ESP32 in deepsleep? All I've done is a deepsleep time of 30 minutes ... if it's more than 30 minutes it's not waking up again ffrom deepsleep...

**Comment by etimou** [comment]
I've run the deep sleep example for more a little bit more than one hour
with no problem. It think the counter was around 800.

On Thu, Nov 16, 2017 at 9:48 AM, cristianhumelnicu <notifications@github.com

**Comment by cristianhumelnicu** [comment]
@etimou Do you use the deepsleep example from Arduino or other scatch ?
Thk!

**Comment by mikey60** [comment]
I am using my own program based on the Arduino deepsleep example an it sleeps for 20 minutes and then sends sensor data to ThingSpeak.  It has run for several days.  When I get a chance I will change the deepsleep to one hour and see if it continues to work.

**Comment by iotmodules** [comment]
I'm using the Sparkfun esp32 thing and am having similar issues however its not just sleep, i'm getting hanged reboots when pressing the reset button too, not just in the TimerWakeUp sketch. 

I also find that I get about 5 reboots before a hang when RTC_DATA_ATTR int bootCount = 0; is intialised however if I comment it out and all references to it then I get about 30-40 reboots before a hang. 

Using latest  with 80a7d60 commited 6 days ago

**Comment by CelliesProjects** [comment]
Using the MHET Live D1 mini-kit.
Can confirm the problem with hanging reboots.

**Comment by etimou** [comment]
I am using the deep sleep example from Arduino, not modified.

**Comment by euquiq** [comment]
I'm having the same problem (with latest Arduino - idf):

 I set up a deep-sleep / wake timer for N seconds (5, 2 or even 1 second) and after a low number of wake-ups (2 to 6 times of normal waking up), ESP32 hangs.

So after reading the above posts, I am waiting for the bug to be crushed (on IDF, it seems).

Regards,

Enrique.

**Comment by copercini** [comment]
There is a commit that fixes it on IDF https://github.com/espressif/esp-idf/commit/abacf8d2a0347a4bf95315cdb227e9b8b1c93794 seems like just wait @me-no-dev bump the version for Arduino IDE  =)

**Comment by me-no-dev** [comment]
need to check if @nkolban has updated the BLE lib and find some time to put a new release.

**Comment by nkolban** [comment]
That is job number 1 starting now

**Comment by cristianhumelnicu** [comment]
@me-no-dev  Could you estimate when any fix update for deepsleep and reset  issue will be availble ?

**Comment by me-no-dev** [comment]
I have updated Arduino. Can you pull the latest changes and give it a try?

**Comment by cristianhumelnicu** [comment]
@me-no-dev I already did .. using deepsleep example with timer wake-up and put to be 60 min in deep sleep . Until now have 12 deepsleep cycles without problems for 1h deepsleep

But i still have one problem .. I put another 3 Wroom32 (Rev.1) to sleep 2h , 6h and 8h ant all of them standing in sleep about 45 minutes and after that wake up . All i've done with success is 1h sleep time with first Wroom32 

Could u guide me what is supouse to do for long time deepsleep i mean 8h - 24h ?

Thnak you very much!

**Comment by me-no-dev** [comment]
@cristianhumelnicu please post a minimal sketch that shows that issue and I will run it here on a few boards to see the result

**Comment by cristianhumelnicu** [comment]
@me-no-dev Is the standard deepsleep example and i use Wroom32 Silicon Rev.1 chipset

#define uS_TO_S_FACTOR 1000000  
#define TIME_TO_SLEEP  7200        

RTC_DATA_ATTR int bootCount = 0;

void setup(){
  Serial.begin(115200);
  delay(200); 

  ++bootCount;
  Serial.println("Boot number: " + String(bootCount));
 
  esp_sleep_enable_timer_wakeup(TIME_TO_SLEEP * uS_TO_S_FACTOR);
  Serial.println("Setup ESP32 to sleep for every " + String(TIME_TO_SLEEP) + " Seconds");

 Serial.println("Going to sleep now");
  esp_deep_sleep_start();  
}
void loop(){ 
}

After last update is working for 1h deep sleep, but if I set TIME_TO_SLEEP>2h , the microconttroler exit from deep sleep in 45min

Until last arduino-esp update  never did exit from deepsleep  for TIME_TO_SLEEP > 30 min 

I did try to define  TIME_TO_SLEEP as unsigned long with same results

Thank you !

**Comment by me-no-dev** [comment]
OH!!!! I got it ;)
```(TIME_TO_SLEEP * uS_TO_S_FACTOR)``` this will default to 32 bit integer as opposed to 64bit one (which is what the wakeup timer expects), so in essence you overflow the 32bit number.
Try ```((uint64_t)(TIME_TO_SLEEP) * uS_TO_S_FACTOR)``` instead

**Comment by euquiq** [comment]
Hi there ! I got a slightly different issue: I got a wakeup timer, which at this time is set for 3 seconds. When the ESP32 awakes, it produces two beeps in a piezoelectric, then goes to sleep.

After about 2 - 3 hours of working about OK, timing goes wrong and it starts to wake up each 1 second by itself. 

At first I had a screen connected thru SPI to the ESP32 showing battery level when awake, but the screen started to be unresponsive at the time of the "acceleration". So I took it off the system just in case it was something related to it. 

It seems to be not. Bottom line, after some time, timer awakening seems to default to another timing. 

Right now I am in the process of creating a minimal example code and reproduce the problem with it. I will post it ASAP.

Regards,

Enrique.

**Comment by mikey60** [comment]
I apologize for the late response.  I actually cloned the esp32-arduino
version 12 hours ago and my sketch worked just fine overnight. It has
executed flawlessly for the past 12 hours or about 36 twenty minute
deep-sleep periods.  Also this version of esp32-arduino seems to have
corrected the Arduino OTA issue I was experiencing with the ESP32 Radio
sketch.   Thank you for all the hard work.

**Comment by fiskfan** [comment]
I pulled the latest release now as well and it seems as if the device starts correctly after deepsleep now.

I would also like to say that it's boot is better overall now. Previously it could also fail to load the sketch after i pressed the reset button, but now it seems as if it always successfully starts the sketch.

**Comment by cristianhumelnicu** [comment]
Well i confirm to  .. the deepsleep and reset is running normal after latest arduino-esp32 update

I think in this moment we can close the issue !

@me-no-dev  Thank you for all !

**Comment by euquiq** [comment]
@cristianhumelnicu not so fast ... ! I am trying this code:

```
#include <esp_deep_sleep.h>

//----Defines 
#define uS_TO_S_FACTOR 1000000  /* Conversion factor for micro seconds to seconds */
#define SLEEP  3 //3 seconds
#define pinBeeper 32

//RTC_DATA_ATTR int bootCount = 0;

void setup()
{
	ledcSetup(0, 2000, 8); //for tone generation
	ledcAttachPin(pinBeeper, 0); //Beeper GPIO to channel 0
	tone(5000, 50); //Happy tone
	tone(6000, 50);
	esp_sleep_enable_timer_wakeup(SLEEP * uS_TO_S_FACTOR);
	esp_deep_sleep_start();
}


void happytone() {
	tone(5000, 50); 
	tone(6000, 50); 
}

//Genero un tono
void tone(int freq, int dur) {
	ledcWrite(0, 127); 
	ledcWriteTone(0, freq);
	delay(dur);
	ledcWrite(0, 0);
}

void loop()
{
	//This is not going to be called
}
```

It starts OK, the beeping occurs on the correct timing. BUT after a couple of hours I cronometrize it again, and instead of 3 seconds, on this opportunity the beeps are spaced about 4 seconds and a half. 

Today in the morning, a similar test, made the beeping go from the programmed 3 seconds down into 1 second. 

I would say that something WEIRD is going on ... maybe it is my particular ESP32 board, so I will try now with a newer board, which also is version 1 silicon inside. (The one I am trying now is Version 0).

**Comment by euquiq** [comment]
Maybe it is better for me to open another issue ?

**Comment by cristianhumelnicu** [comment]
@euquiq 

#warning esp_deep_sleep.h will be deprecated in the next release. Use esp_sleep.h instead.

**Comment by euquiq** [comment]
@cristianhumelnicu thanks for the EYESUP on the deprecation for esp_deep_sleep.h ...

On the delay(3000) issue, this is part of a burn - in test for a system that operates under solar panel / supercapacitors system so I really need to put the ESP32 to sleep.

 I find weird that sleep time may go faster or slower than the programmed value. Sometimes for a significant %. I am at this time performing tests on different sleep times, to see it's behavior.

 I do not think it is "Boot time" related, not at differences drifting from 3 seconds down to 1 second or 4.x seconds, when WIFI is off, etc. 

But if you can throw any idea or light into it, I will try to understand / process it into my tests.

Thank you, Enrique

**Comment by degehi** [comment]
I tried the last revision, like @mikey60 @fiskfan and @cristianhumelnicu, those last days and I still have reboot problems. It seems better than before (I use ESP32 with a 1 hour deepsleep and it can run a day without any problem) but finally, sooner or later, it will hang again on wake-up. I tried less than 1 hour deepsleep but same results

If I come back with version 9512368, it works fine even after several days.

What I use:

```
#define uS_TO_S_FACTOR 1000000  /* Conversion factor for micro seconds to seconds */
#define TIME_TO_SLEEP  3600     /* Time ESP32 will go to sleep (in seconds) */
...
// init deep sleep
esp_sleep_enable_timer_wakeup((uint64_t)(TIME_TO_SLEEP) * uS_TO_S_FACTOR);
...
esp_deep_sleep_start();
```
Hardware: Sparkfun ESP32 Thing

Nobody has the same problem than I?

**Comment by mvadu** [comment]
I can see the same problem on a Adafruit Huzzah32 feather board. I have a simple modified deep sleep sketch which logs the booth count to a influxdb (which is plotted with grafana). Sleep duration is 120 seconds. As you can see below it randomly gets rest to 0. Max I have seen it grow is 700 times (close to a day). 

![image](https://user-images.githubusercontent.com/8618235/34342697-babb80a6-e985-11e7-817c-84b7a800ab6f.png)

**Comment by cloudiaworld** [comment]
Any news regarding this issue? I continue having the same issue on some boards; namely that the esp32 stops waking-up correctly after some (non-deterministic) cycles of sleep/wake up.  I get, in addition to the DEEPSLEEP_RESET a TG0WDT_SYS_RESET.  All the values I have stored as RTC_DATA_ATTR are gone after the problem occurs. 

 Any hints? I am using a ESP-WROOM-32 module. 

Core Installation/update date:	08/02/2018
IDE name:	Arduino IDE 1.8.5
Flash Frequency:	40Mhz / also happens at 80MHz
Upload Speed:	115200

rst:0x5 (DEEPSLEEP_RESET),boot:0x12 (SPI_FAST_FLASH_BOOT)
ets Jun  8 2016 00:22:57

rst:0x7 (TG0WDT_SYS_RESET),boot:0x12 (SPI_FAST_FLASH_BOOT)
ets Jun  8 2016 00:22:57

rst:0x7 (TG0WDT_SYS_RESET),boot:0x12 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0018,len:4
load:0x3fff001c,len:956
load:0x40078000,len:0
load:0x40078000,len:13076
entry 0x40078a58

**Comment by misan** [comment]
Same problem here. It works for a while but eventually does not wake up. Or more exactly, it hangs on wake-up, as the fact the chip is hot makes me think it is no longer in deep-sleep. 

For low-power IoT applications (as it is mine) this issue is a deal-breaker.

**Comment by wagnerfl85** [comment]
Same problem here, i also use a ESP32 Dev Module and it get stuck after some time. My sketch runs as follows:

1. wakeup
2. connect to wifi
3. get data from BME280
4. send data via mqtt to openhab
5. go to deep_sleep
6. wakeup after 5 Minutes

The last run was about 30-35 hours without a problem but then stucks again. After a button reset it doe work again.

**Comment by gianky69** [comment]
I have a similar sketch like @wagnerfl85 (ESP32, wifi, BME data...) . I send data to Home Assistant via MQTT.
Using latest Arduino build the  ESP32 battery module hangs after 2-3 hours!

**Comment by mkihr** [comment]
Same issue here. With both adafruit feather and WEMOS ESP32. Hangup when wakeup from deep sleep.  I only tried with 1 to 10 minutes. After some wakeup, it stuck. Can be after 10 or after 700.
As I see here this is an ongoing issue for long time. Will there be a solution?

**Comment by axelisart** [comment]
Hi everybody. I don't know if it can help but here are some observations again : I tried with a raw esp32 (just the chip, not mount in any dev board) and I am near 1 month without any stall. It represent 3786 wakes as it wakes every 10mn. The esp32 is just connected to an SHT31 sensor to get temp/hdty. It is powered through a tsp-03 step down buck (220v to 3,3v). I have lot of different ESP32 and ESP8266 Chinese dev boards working in parallel in the same wake up frequency and every of them stalled lot of times during the same period. So for the moment my conclusion is that it is maybe not due to the ESP chip itself but more to the board it is attached. So if you absolutely need to run a project with ESP, look a my workaround.  I used a Board Plate Pinboard Convertor to solder the esp32 (50 cents) and an ESP-WROOM32 (3$). I just soldered required pins for my sht sensor and tx/rx to program it. It is just a bit difficult to solder chip pin to the board but it is feasible. I will keep posting here if I see any stall for my "naked" ESP32.

**Comment by misan** [comment]
That is a very good point @axelisart I am going to monitor what happens to the power supply during the wake-up from sleep. It may well be that many of the existing boards are not properly design to handle well this sudden load change. Thanks for reporting it, as many of us are trying to figure out why we could not get a reliable result when apparently others have no complaints.

**Comment by misan** [comment]
No wonder there is wake-up trouble, supply rail looks like sh*t on restart. Next stop is to get a decent 3v3 supply.

![image](https://user-images.githubusercontent.com/466120/46160660-a506f100-c283-11e8-9787-4e24bdf1f183.png)

**Comment by stickbreaker** [comment]
@misan I see those droops (5 I think) are off scale, How far down does they go?

Chuck.

**Comment by misan** [comment]
It appears they go almost 1Volt down.

![image](https://user-images.githubusercontent.com/466120/46161425-d2549e80-c285-11e8-9261-00cc9c0d6a4a.png)

**Comment by stickbreaker** [comment]
@misan Wow, that regulator is useless, If I am reading you scope correctly, 800mV droop, 800mV overshoot.  So the 3.3V drops down to 2.5V.  A 24% reduction! and then jumps to 4.1V on the rebound.  That is ridiculous.

Chuck.

**Comment by stickbreaker** [comment]
@misan It's almost like you have an inductor on the power line. 

Try changing your USB cable,  maybe you have a one with a builtin choke.

Chuck.

**Comment by misan** [comment]
@stickbreaker Thanks Chuck, similar results powering it from Raspberry PI 2A power brick. AMS117 LDO. I'll try a larger capacitor after the regulator.

**Comment by stickbreaker** [comment]
The specs for a AMS1117 show only a 25mV variance from 0 to 800mA load, but that is with input held at 4.75V.
![ams1117 load regulation](https://user-images.githubusercontent.com/2858309/46162880-cac1d500-c246-11e8-84b7-9ecc7d22ac8d.JPG)

You are definitely going to have to beef up that power circuit.

Chuck.

**Comment by stickbreaker** [comment]
@misan 
It could even be board layout issues. Sometimes there is no consideration for instantaneous current flow when traces are laid out.  a 10mil trace can't handle much power!

Chuck.

**Comment by misan** [comment]
@stickbreaker good point, maybe a tricker wire is all that is needed. I will keep on searching ,thanks a lot.

**Comment by blackbrayn** [comment]
Having similar issues trying to deep sleep a wemos lolin32 v1 clone board, after a few cycles of deep sleep it just prints dots in the serial , left the esp connected on the laptop for serial monitoring.
Code also doesn't run , i see no connections to the server it sends the data.

**Comment by chegewara** [comment]
This week ive been doing test with this example and it worked for 30-40 cycles with no issue (short test): https://github.com/nkolban/ESP32_BLE_Arduino/blob/7951347ed68313d75c367e1f2cce763cb56d1eb2/examples/BLE_iBeacon/BLE_iBeacon.ino

**Comment by misan** [comment]
@blackbrayn Not sure about your code, but many programs will print dots while WiFi access is being secured. If that is the case, it may well be printing forever is access is never granted for any reason.

That was one of the problems I experienced that looked like a fault of the deep sleep but it was actually an endless loop in my code once Wifi authentication was not successful. After a closer consideration, it was user error (me) and not the ESP fault. But the fact that data logging stopped at a given point looked like the chip was at fault.

That said, I can easily trigger a fault condition in the TimerWakeUp example if I commented out the delay in

`void setup(){

Serial.begin(115200);

delay(1000); //Take some time to open up the Serial Monitor`

It is my understanding that delay helps getting things settled, as it will fail after a while if commented out. Also, some boards will change the serial port every time I upload new code (while others do not do that).

**Comment by blackbrayn** [comment]
@misan Thx for the tips , i'll increase the time to see if it helps(thinking it's maybe too agressive for the wireless).
Just noticed that running on battery seems to be more stable for now (maybe the voltage doesn't drop that much) , i'll do some more tests on the electric side once i get some time , maybe solder a capacitor right to the pins of the board that conect directly to the esp module.
Running the battery thru the multimeter on the low amps side(100mA) causes instant problems (no boot) i supose , using the 10 amp range that has a beefy shunt and a lower resistance works fine but can't really measure deep sleep curent.

**Comment by mvadu** [comment]
I always had it running on a lipo battery, and still had the problem. Not sure about the spoke dropping voltage resulting in a reset.

Regards,
Adarsha

sent from handheld, excuse typos.

________________________________

**Comment by blackbrayn** [comment]
Just ran a test on lipo battery with no other modifications(just a multimeter in series) to really see if it's more stable on my board , it just stopped after a while, it worked for only one hour and after that it started drawing 120mA all the time.
![img_20181004_163953](https://user-images.githubusercontent.com/43824644/46478349-72b24200-c7f5-11e8-93a0-4e29b7934a11.jpg)

**Comment by blackbrayn** [comment]
@StefanStsc Hello , i did not continue the power related tests , i managed to get  better stability (but not fully stable) by adding sleep timers before and after enabling wifi , and  before and after the  other operations in the code.
I also got a barebone esp32 module to test with a diferent power suply and the mininmum conections.

**Comment by JLIspace** [comment]
I'm also facing a similar (the same?) issue. My ESP32 randomly stops after waking up. I tested it a few times. The result was:
Stuck @ 64. boot
Stuck @ 54. boot
Stuck @ 132. boot
While only deep sleeping for 15seconds in between. I used delay(1000) at the beginning of the setup.
I can't really tell what's going on. There are no signs of under-volting.

Any suggestions? :)

**Comment by muczy** [comment]
The only thing helped in my case (ESP32 sending readings from BME280 via MQTT, then sleeping for 5 minutes, operated using a lifepo4 battery, no response after ~1.5 days) was a big fat capacitor...

**Comment by JLIspace** [comment]
> The only thing helped in my case (ESP32 sending readings from BME280 via MQTT, then sleeping for 5 minutes, operated using a lifepo4 battery, no response after ~1.5 days) was a big fat capacitor...

I already have a pretty big one on my 5V supply rail. There are some other compenents using it but I haven't seen a significant drop yet. (lowest was about 4,9V)
Could it be that the 3V3 voltage regulator built into my dev board can't handle the esp itself? 🤔
Available power shouldn't be an issue as it's directly comming from a rather big LiFePO4 battery pack.

I previously had problems with weak power supplies causing the esp32 to crash and restart but this is the first time that it just quietly quits doing its job.
I'm basically planning to run this setup 24/7 so having it fail after as low as 5 minutes is a major flaw :D

Thanks for your suggestion! I'll definately verify the voltage again.

**Comment by krish2487** [comment]
Hello, I would like to pitch in with my experience. I too am facing a similar issue. The board in question is a custom ESP32 wrover board powered from a LD1117 3.3V regulator with a  10uF input and output capacitance. As suggested a couple of times on the thread, I did place a beefy (1000 uF/ 10V) elcap directly across the ESP32 power pins with short leads. 

However, I still get the same stalling issue when the device reboots and halts without executing any code. The code is more or less this, wake up, read from BME680 and TSL2591, send data. wait till next transmission time and repeat. If the device is not able to post to the server or get an IP at any point of time, the device reboots.

I would be extremely grateful for any other suggestions to fix this issue. :-)

Thanks in advance!!

**Comment by blackbrayn** [comment]
@krish2487 hello , did you try adding wait timers before enabling wifi or conecting to server? Maybe it helps a bit like in my case , not a permanent fix but it's something.

**Comment by krish2487** [comment]
@blackbrayn  I did not explicitly add any other delay. There is already an existing delay in the firmware.
This is the specific bit for wifi connection.

```
        WiFi.disconnect();
        delay(500);
        wifiConnected = false;
        WiFi.begin(ssid, password);

```

I am assuming that thats what you meant by adding a wait timer? There is already a half a second delay. Should I increase it??

Thanks for the suggestion. Anything that helps is appreciated. :-)

**Comment by cookie0815** [comment]
Had the same issue, but also with permanently looping functions. 2 to 12 hours of operation, then silence.
Very stable power supply solved my problems - in this case a 1000µF Capacitor between 3,3V and GND, as well as thick wires, solded directly to the ESP32 Development Board.
(I'm powering through the 3,3V Pin from a small power supply that also powers two "rails" of pins for several sensors. The rails already had a capacitor and the cable to the ESP was only about 8cm long, but in the end I needed the measures mentioned above - well, maybe not all of them, I installed them all at once, but the package works...).
So, power supply seems crucial, see also https://arduino-hannover.de/2018/07/25/die-tuecken-der-esp32-stromversorgung/ (in German though).

Talking software - I use a 500ms delay first, to allow proper wakeup. And also some delay at the end, to allow proper MQTT publishing of sensor data (seemed to be necessary after the first tests).

**Comment by blackbrayn** [comment]
> Had the same issue, but also with permanently looping functions. 2 to 12 hours of operation, then silence.
> Very stable power supply solved my problems - in this case a 1000µF Capacitor between 3,3V and GND, as well as thick wires, solded directly to the ESP32 Development Board.
> (I'm powering through the 3,3V Pin from a small power supply that also powers two "rails" of pins for several sensors. The rails already had a capacitor and the cable to the ESP was only about 8cm long, but in the end I needed the measures mentioned above - well, maybe not all of them, I installed them all at once, but the package works...).
> So, power supply seems crucial, see also https://arduino-hannover.de/2018/07/25/die-tuecken-der-esp32-stromversorgung/ (in German though).
> 
> Talking software - I use a 500ms delay first, to allow proper wakeup. And also some delay at the end, to allow proper MQTT publishing of sensor data (seemed to be necessary after the first tests).

Excellent article , covers problematic scenarios when powering the esp module with actual measurements , i will try to do some more testing on my module as well when i get time.

**Comment by alexvancasper** [comment]
> Very stable power supply solved my problems - in this case a 1000µF Capacitor between 3,3V and GND, as well as thick wires, solded directly to the ESP32 Development Board.

What was a voltage of this capacitor? I have 16V, is it fine or not?

**Comment by blackbrayn** [comment]
Higher voltage on the filtering capacitor will not be a issue , except larger dimensions there is no problem.

**Comment by kenken64** [comment]
I am using a micro usb cable the outcome still the same

**Comment by blackbrayn** [comment]
More info and related issues regarding stability when powering wireless :
https://github.com/esp8266/Arduino/issues/2070

I finally managed to get stable(1 month and running) deep sleep (timer 5 minutes) on the lolin32 with a 18650 battery conected to the board by switching code for transmitting data to http post and a 1000ms delay after Serial.begin and WiFi.begin.
Exact code used (didn't have time to clean it up, just changed the domain and wifi credentials) :
https://github.com/blackbrayn/Lolin32-Deep-Sleep
I also put the wireless network down and the server receiving the data to be sure the code loop won't freak out . It just proceeds to sleep& try to send data. (i only adjusted the wireless connect a bit , the code is commented).

**Comment by StefanStsc** [comment]
Because the bot wants to close it: this problem is still happening.

**Comment by basil59** [comment]
Sorry to jump onto a 2 year old thread, but I'm having the same issue.
Any progress? anything I can provide to help with troubleshooting?
thanks.

**Comment by StefanStsc** [comment]
This seems to happen if the power supply is not perfect during power on. I could fix it with some capacitors. Anyway I consider this as a workaround, because the same power supply does not result in any problems during normal operation.
So no real progress and no real fix.

**Comment by StefanStsc** [comment]
Bot, forget that. Get it fixed.

**Comment by StefanStsc** [comment]
Still not solved.

**Comment by JLIspace** [comment]
Yeah, it's still happening for me.

**Comment by bglowacki** [comment]
Got the same issue.

**Comment by alexisartus** [comment]
Also still the same problem

**Comment by StefanStsc** [comment]
And I've also this issue for more than one year.

**Comment by bglowacki** [comment]
Hmm I removed code that reads from Capacitive Soil Moisture sensor, and disconnected the sensor from the ESP32 and it seems like the issue doesn't happen for now. I have been successfully running the DeepSleep cycles for almost a day.

Going to reconnect the sensor now and see what happens.

**Comment by bglowacki** [comment]
OK, seems like it is working don't know the actual thing that caused the problem.
The code that had problem was

```
#include <Arduino.h>
#include <WiFi.h>
#include <PubSubClient.h>
#include <ArduinoJson.h>

// IPAddress ip(192,168,1,172);
const char* topic = "test_sensor_2";
// #define SLEEP_TIME 3600000000 // 1 hour
#define SLEEP_TIME  600000000 // 10 minutes
// #define SLEEP_TIME  60000000; // 1 minute


const char* ssid = "*******";
const char* password = "********";

WiFiClient espClient;
PubSubClient client(espClient);

const char* mqtt_server = "192.168.1.128";

const int analog_pin = 36;
StaticJsonDocument<256> doc;

// Ticker sleep_ticker;

void connectWiFi() {
  // IPAddress gateway(192,168,1,254);
  // IPAddress subnet(255,255,255,0);  
  
  delay(1);
  WiFi.persistent(false);
  WiFi.mode(WIFI_STA);
  // WiFi.config(ip, gateway, subnet);
  WiFi.begin(ssid, password);

  Serial.print("Connecting");
  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
      Serial.print(".");
  }
  Serial.println();

  Serial.print("Connected, IP address: ");
  Serial.println(WiFi.localIP());
}

void connectMqtt() {
  while (!client.connected()) {
  Serial.print("Attempting MQTT connection...");
  client.setServer(mqtt_server, 1883); 
  if (client.connect("ESP8266 Client")) {
    Serial.println("connected");
    } 
  }
}

void getReadings(int numberOfReads) {
  int soil_moisture_value = 0;
  int soil_moisture_percent = 0;
  const int air_value = 3104;
  const int water_value = 1403;

  for (int i=0;i<numberOfReads;i++) {
    soil_moisture_value = analogRead(analog_pin);
    delay(100);
  }

  soil_moisture_percent = map(soil_moisture_value, air_value, water_value, 0, 100);

  if(soil_moisture_percent > 100) {
    soil_moisture_percent = 100;  
  } else if(soil_moisture_percent <0) {
    soil_moisture_percent = 0;
  }
  
  doc["sensor"] = "soil moisture";
  doc["raw_value"] = soil_moisture_value;
  doc["value"] = soil_moisture_percent;
  doc["rssi"] = WiFi.RSSI();

}

void publishMessage() {
  char buffer[256];
  size_t n = serializeJson(doc, buffer);
  bool published = client.publish(topic, buffer, n);
  delay(500);

  Serial.print("published: ");
  Serial.println(String(published));

  Serial.println(buffer);
  client.disconnect();
  delay(500);
}

void sleep() {
  esp_sleep_enable_timer_wakeup(SLEEP_TIME);
  WiFi.disconnect();
  delay(500);
  esp_deep_sleep_start();
 }

void setup() {
  Serial.begin(115200);
  // sleep_ticker.once_ms(200000, &sleep);
  connectWiFi();
  connectMqtt();
  getReadings(5);
  publishMessage();
  sleep();
}

void loop() {
  // put your main code here, to run repeatedly:
}
```

The code that does not have problem is:

```
#include <Arduino.h>
#include <WiFi.h>
#include <PubSubClient.h>
#include <ArduinoJson.h>

// IPAddress ip(192,168,1,172);
const char* topic = "soil_humidity_sensor_1";
// #define SLEEP_TIME 3600000000 // 1 hour
#define SLEEP_TIME  600000000 // 10 minutes
// #define SLEEP_TIME  60000000; // 1 minute


const char* ssid = "*********";
const char* password = "**********";

WiFiClient espClient;
PubSubClient client(espClient);

const char* mqtt_server = "192.168.1.128";

void setup() {

  ////////// Connect WiFi
  WiFi.begin(ssid, password);
  Serial.print("Connecting");
  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
      Serial.print(".");
  }
  Serial.println();

  Serial.print("Connected, IP address: ");
  Serial.println(WiFi.localIP());

  ////////// Connect MQTT
  while (!client.connected()) {
  Serial.print("Attempting MQTT connection...");
  client.setServer(mqtt_server, 1883); 
  if (client.connect("ESP8266 Client")) {
    Serial.println("connected");
    } 
  }

  ////////// Get Readings

  int soil_moisture_value = 0;
  int soil_moisture_percent = 0;
  const int air_value = 3104;
  const int water_value = 1403;

  for (int i=0;i<5;i++) {
    soil_moisture_value = analogRead(34);
    delay(100);
  }

  soil_moisture_percent = map(soil_moisture_value, air_value, water_value, 0, 100);

  if(soil_moisture_percent > 100) {
    soil_moisture_percent = 100;  
  } else if(soil_moisture_percent <0) {
    soil_moisture_percent = 0;
  }

  ////////// Publish Message
  char message[16];
  itoa(soil_moisture_percent, message, 10);
  bool published = client.publish(topic, message);
  delay(500);

  Serial.print("published: ");
  Serial.println(String(published));
  delay(500);

  ////////// Go to Sleep
  esp_sleep_enable_timer_wakeup(SLEEP_TIME);
  delay(1);
  esp_deep_sleep_start();
}

void loop() {

}
```

**Comment by rmhomecouk** [comment]
Regrettably I’m suffering from this for a project that requires deep sleep - my unit is a tinypico with a stable power supply and a solar panel to keep it charged. It’s very random on how long it will run with deep sleep but it always fails eventually requiring physical access/reset

**Comment by mvadu** [comment]
@rmhomecouk since you have a backup power may be you can trigger restart at certain intervals and test if the need for physical access goes away..

**Comment by JLIspace** [comment]
I went with using the watchdog timer for resetting the esp32 once locked up. Periodically resetting it did not work for me. It seems like the issue has a totally sporadic cause...

**Comment by fbambusi** [comment]
I am experiencing the same problem. @StefanStsc, could you share a schematic with the capacitors you used to address the problem?

**Comment by iotmodules** [comment]
Try using a voltage supervisor on the en pin.

**Comment by fbambusi** [comment]
Hi, unfortunately, the voltage supervisor did not solve the problem. However, light sleep does not seem to have the same problem.
Two caveats in case you want to adopt this solution:
- Light sleep's current consumption is 0.8 ma, compared to 10ua of deep sleep.
- When the ESP32 awakes from light sleep, it resumes the execution after the esp_light_sleep_start() instruction: therefore, in case you relied on esp_deep_sleep_start() to restart the program, you should add extra code to reproduce this behaviour.

**Comment by StefanStsc** [comment]
@fbambusi I can't remember exactly. Too long ago. It was capacitors on 3V3 pin and separate one on EN. Took some time to find the correct combination but worked.
Anyway, I still insist that this is a bug that needs to be solved.

**Comment by pavramucz** [comment]
As a note: I saw this problem when writing Arduino code by hand, but can no longer see it since I moved to Esphome. I know Esphome might not be an alternative, for quite some, but might help in checking if you have a hardware issue.

**Comment by NaimSakaamini** [comment]
facing a similar issue, it works fine for around 200 cycles then fails to wake up again.

**Comment by iotmodules** [comment]
Ok there is an option to delay the startup of the processor to give the flash time to initialise. Not so easy to configure using an Arduino build but with esp-idf you can set it with makemenu config.

**Comment by sagar448** [comment]
@iotmodules Did you figure it out? So you're saying delaying the startup of the processor fixed it?

**Comment by sagar448** [comment]
@JLIspace Did the watchdog timer fix the issue for you?

**Comment by StefanStsc** [comment]
@sagar448 As a workaround I was using the watchdog timer. It worked with ESP IDF. It was not an ideal solution because it increased the power consumption a lot.

**Comment by iotmodules** [comment]
https://youtu.be/cKDv0aN67BY

This is a very good explanation of how to use voltage supervisors to make sure the esp32 boots properly.

**Comment by NaimSakaamini** [comment]
> Ok there is an option to delay the startup of the processor to give the flash time to initialise. Not so easy to configure using an Arduino build but with esp-idf you can set it with makemenu config.
> […](#)
> Sent from my iPhone
> On 11 Aug 2020, at 15:56, Naim Sakaamini ***@***.***> wrote: ﻿ facing a similar issue, it works fine for around 200 cycles then fails to wake up again. — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or unsubscribe.

I'm trying to add the delay. Where in the menu config were you able to set it up?
thank you

**Comment by lbernstone** [comment]
CONFIG_ESP32_DEEP_SLEEP_WAKEUP_DELAY

**Comment by NaimSakaamini** [comment]
Thank you @lbernstone. I had it already set to 2000ms. 
You think I should increase it to avoid the issue, or is 2000ms sufficient enough?

**Comment by lbernstone** [comment]
2000 is the default in arduino.  If you are seeing `flash read err, 1000` in your logs, then you should increase it.  If you have no logs, then you are just chasing ghosts.

**Comment by NaimSakaamini** [comment]
I have no logs. It only goes to sleep after random number of cycles and never wakes up. Am I supposed to do something to get the logs ?

**Comment by Kashyapkoshti** [comment]
rst:0x10 (RTCWDT_RTC_RESET),boot:0x1f (SPI_FAST_FLASH_BOOT)
flash read err, 1000
ets_main.c 371 

I am getting this error in my esp32 devkit v1
How to solve it?
I am using arduino IDE for programming the esp32

**Comment by lbernstone** [comment]
Please don't hijack issues.  Your problem clearly has nothing to do with deep sleep.

**Comment by Kashyapkoshti** [comment]
> Please don't hijack issues. Your problem clearly has nothing to do with deep sleep.

I am not hijacking your issue,i dont have any clue about my error.
 i read somewhere in your comments that flash read errr,1000 and i thought maybe this is related.

**Comment by StefanStsc** [comment]
Not solved

**Comment by zeekoe** [comment]
I'm quite sure I have a similar issue, on MicroPython. Using deep sleep, sometimes my ESP32 (Heltec Wifi kit 32) doesn't wake up. It runs on 4x NiMH + a diode to lower the voltage. Works perfectly (it's not charging the batteries of course). When the battery voltate is lower, it fails to wake up significantly more often. So I can confirm the feeling of the battery being related.
MicroPython doesn't support other sleep modes, so I cannot test them. I assumed max watchdog timer value to be seconds-ish, but @JLIspace 's comment suggests it can be longer; @JLIspace could you elaborate just a little bit on your solution?

**Comment by AIexBV** [comment]
I am running some Wroom32 boards to send BME680 values to a MQTT server every 15 minutes. The circuits are powered by 2 AA batteries (3 volts). Some of the boards run flawlessly over months, while some of them fail after some days and drain the battery within hours. My latest attempt soldering a capacitor of 100uF very close to the board to filter the power supply seems to be successful. All of them but one are running now for 2 weeks without a problem. At one board the capacitor did not bring the solution. I changed the capacitor to a 47uF Tantal today. Maybe that works. Basically I guess the stability of the voltage is the key solution for the problem.

**Comment by Younes-SadatNejad** [comment]
Hello
I am having the exact same problem. 
I am using ESP32 Dev kit1, with the external trigger as my wake-up from a deep sleep. 
The problem is that after few wakeup and sleep, it stops waking up and manual reboot is the only way to fix it. 
The problem exists both with wire and without wire, however when using battery, it happens sooner (about 30-40 times), while with wire it may take 200 times. 
I used a pull-up resistor of 100K for pin2 that is my trigger, and a 100uF capacitor on the battery, the problem remains. 
I tried adding a 500ms delay at the beginning of setup, the problem remains. 
Is there any software solution, or is it just a matter of finding the right capacitor on the voltage supply?

**Comment by iotmodules** [comment]
If you are using the esp-idf try leaving the RTC slow memory enabled when configuring deep sleep.

**Comment by Waged** [comment]
I have the same issue too with ESP32-WROVER the wake-up after deep sleep and it seems everything is working fine for the flash read   and here us the

`ets Jun  8 2016 00:22:57

rst:0x5 (DEEPSLEEP_RESET),boot:0x33 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0018,len:4
load:0x3fff001c,len:1044
load:0x40078000,len:8896
load:0x40080400,len:5816
entry 0x400806ac`

i was testing the Voltage and the pin status of EN and during touching the En pin to check 3.3 voltage is present and just with one probe "may be static charge" the ESP restarted and worked fine and the restart reason was waking up from deep_sleep although it stuck for hours and the sleep was just for 5 minutes. any ideas?

**Comment by ppotter100** [comment]
A possible fix is mentioned in https://github.com/espressif/esp-idf/issues/6694 with batteries.

**Comment by ginixsan** [comment]
Tried all possible solutions here and still same problem. It wakes up from deep sleep and stalls in setup after 2-3 lines.. no cause, no reset, no crash.. nothing.. just freezes...

**Comment by JLIspace** [comment]
> Tried all possible solutions here and still same problem. It wakes up from deep sleep and stalls in setup after 2-3 lines.. no cause, no reset, no crash.. nothing.. just freezes...happens with batteries and with usb connected..

Yep. My bet is that a bad power supply still is the main cause. Using a watchdog timer is the only method of recovering after this „freeze“ that I‘m aware of.

best regards JLIspace

**Comment by oktett-8** [comment]
> Tried all possible solutions here and still same problem. It wakes up from deep sleep and stalls in setup after 2-3 lines.. no cause, no reset, no crash.. nothing.. just freezes...happens with batteries and with usb connected..

Also was going mad with this issue for years, excluded power supply issues and using different chip revisions (last tests done with ESP32-E) and didn’t find a root cause, but a least this kind of workaround for improvement:
If you do a `delay(100);` right at the beginning of `setup()` most ESP32 show an improvement. Some chips that got stuck before every couple of minutes then ran for years without issues (on battery since this was the purpose of my deep sleep application). Some chips need 500 ms delay to work permanently whilst with others even longer times don’t help at all.

**Comment by epareja2** [comment]
Anyone has found yet the root cause on this?

**Comment by mvadu** [comment]
Around 4 months ago I split my network, so 2.4ghz has a separate ssid. That also means I turned off "Smart Connect" since that is the feature of Asus's code which pushes the clients across 2.4/5ghz bands. Ever since I have had zero disconnects from around 15 tasmota devices.

**Comment by Olfdev** [comment]
I think I'm having the same issue with my AZDelivery lolin32 lite. I thought I was running out of battery even though it didn't make a lot of sense but now I know the battery isn't the culprit. With the USB connected, it hangs forever (after 13 deep sleep/wake up but this is very random), I don't get the "ets Jun  8 2016 00:22:57" message:

`rst:0x5 (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13232
load:0x40080400,len:3028
entry 0x400805e4
ets Jun  8 2016 00:22:57

rst:0x5 (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13232
load:0x40080400,len:3028
entry 0x400805e4
ets Jun  8 2016 00:22:57

rst:0x5 (DEEPSLEEP_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13232
load:0x40080400,len:3028
entry 0x400805e4
`

I'll try the delay at the beginning of the setup and report back.
